lib/engine/locations.ex

```diff
@@ -0,0 +1,150 @@
+defmodule Engine.Locations do
+  @moduledoc """
+  Maintains an up-to-date internal state of the realtime locations of vehicles in the system. Fetches
+  from the GTFS-rt enhanced JSON file about once per second.
+  """
+  @behaviour Engine.LocationsAPI
+
+  use GenServer
+  require Logger
+
+  @type state :: %{
+          last_modified_vehicle_positions: String.t() | nil,
+          vehicle_locations_table: :ets.tab()
+        }
+
+  @vehicle_locations_table :vehicle_locations
+
+  def start_link([]) do
+    GenServer.start_link(__MODULE__, [], name: __MODULE__)
+  end
+
+  @impl true
+  def for_vehicle(locations_table_id \\ @vehicle_locations_table, vehicle_id) do
+    case :ets.lookup(locations_table_id, vehicle_id) do
+      [{_, :none}] -> nil
+      [{^vehicle_id, location}] -> location
+      _ -> nil
+    end
+  end
+
+  @impl true
+  def init(_) do
+    schedule_update(self())
+
+    @vehicle_locations_table =
+      :ets.new(
+        @vehicle_locations_table,
+        [:set, :protected, :named_table, read_concurrency: true]
+      )
+
+    {:ok,
+     %{last_modified_vehicle_positions: nil, vehicle_locations_table: @vehicle_locations_table}}
+  end
+
+  @impl true
+  def handle_info(:update, state) do
+    schedule_update(self())
+
+    last_modified_vehicle_locations =
+      download_and_process_vehicle_locations(
+        state[:last_modified_vehicle_positions],
+        :vehicle_positions_url,
+        state[:vehicle_locations_table]
```

Comment:
nit: could use dot notation on `state` here, and there's no good reason to pass `:vehicle_positions_url` as a parameter, we could just inline it in the function.